crates/core/src/engine.rs

```diff
@@ -1,72 +1,32 @@
 use anyhow::Result;
-#[cfg(feature = "standalone-wasi")]
 use std::io::{copy, stdin, stdout, Write};
 
-#[cfg(not(feature = "standalone-wasi"))]
-#[link(wasm_import_module = "shopify_v1")]
-extern "C" {
-    pub fn input_len(len: *const usize) -> u32;
-    pub fn input_copy(buffer: *mut u8) -> u32;
-    pub fn output_copy(buffer: *const u8, len: usize) -> u32;
-}
-
 pub fn load() -> Result<Vec<u8>> {
-    #[cfg(not(feature = "standalone-wasi"))]
-    return load_from_abi();
-
-    #[cfg(feature = "standalone-wasi")]
-    return load_from_stdin();
-}
-
-#[cfg(feature = "standalone-wasi")]
-fn load_from_stdin() -> Result<Vec<u8>> {
     let mut reader = stdin();
-    let mut writer: Vec<u8> = vec![];
-    copy(&mut reader, &mut writer)?;
-    let value: serde_json::Value = serde_json::from_slice(&writer)?;
-    rmp_serde::to_vec(&value).map_err(Into::into)
-}
+    let mut output: Vec<u8> = vec![];
+    copy(&mut reader, &mut output)?;
 
-#[cfg(not(feature = "standalone-wasi"))]
-fn load_from_abi() -> Result<Vec<u8>> {
-    let len = 0;
-    unsafe {
-        input_len(&len);
+    #[cfg(not(feature = "json-io"))]
+    {
+        Ok(output)
     }
-
-    let mut input_buffer = vec![0; len];
-    unsafe {
-        input_copy(input_buffer.as_mut_ptr());
+    #[cfg(feature = "json-io")]
+    {
+        let output: serde_json::Value = serde_json::from_slice(&output)?;
+        rmp_serde::to_vec(&output).map_err(Into::into)
```

Comment:
Would it be possible to transcode the JSON to msgpack?

Reply:
What's the value of doing that here?

Reply:
Done here https://github.com/Shopify/javy/pull/78/commits/f3a9cf306329fc54657df1bd6579ce2328f37eb9